WSL+Docker: Kubernetes on the Windows Desktop blog post added (#21092)
kubernetes/website: content/en/blog/_posts/2020-05-21-wsl2-dockerdesktop-k8s.md

And in addition, WSL brought an ability to run Kubernetes on Windows almost seamlessly!

Below, we will cover in brief how to install and use various solutions to run Kubernetes locally.

# Prerequisites

As this blog will explain how to install KinD, we won't see in details the installation of the prerequisites.

However, here is the list of the prerequisites needed and their version/lane:

- OS: Windows 10 version 2004, Build 19041
- [WSL2 enabled](https://docs.microsoft.com/en-us/windows/wsl/wsl2-install)
  - In order to install the distros as WSL2 by default, once WSL2 installed, run the command `wsl.exe --set-default-version 2` in Powershell
- WSL2 distro installed from the Windows Store - the distro used is Ubuntu-18.04
- [Docker Desktop for Windows](https://hub.docker.com/editions/community/docker-ce-desktop-windows), stable channel - the version used is 2.2.0.4
- [Optional] Microsoft Terminal installed from the Windows Store
  - Open the Windows store and type "Terminal" in the search, it will be (normally) the first option

>     ![image-20200323160845196](/images/blog/2020-05-21-wsl2-dockerdesktop-k8s/wsl2-windows-store-terminal.png)

Author: @sftim can you give a sample, please?

>     ![image-20200323160845196](/images/blog/2020-05-21-wsl2-dockerdesktop-k8s/wsl2-windows-store-terminal.png)

Reviewer: yes sorry, totally right. I tend to keep it empty, while the name of the picture states what it is.
I will remove the [] content, or for alt, is there a best practice for accessibility? should I be very verbose or just state what it is?

For information, all the pictures are the exact replicas of the commands **before** them

>     ![image-20200323160845196](/images/blog/2020-05-21-wsl2-dockerdesktop-k8s/wsl2-windows-store-terminal.png)

Reviewer 2: You could use “Terminal window screen grab” as the alt text (for example).

>     ![image-20200323160845196](/images/blog/2020-05-21-wsl2-dockerdesktop-k8s/wsl2-windows-store-terminal.png)

Author: Fixed, thanks!

ls $HOME/.kube
# Create the cluster and give it a name (optional)
kind create cluster --name wslkind
# Check if the .kube has been created and populated with files
ls $HOME/.kube
```

![kind-cluster-create](/images/blog/2020-05-21-wsl2-dockerdesktop-k8s/wsl2-kind-cluster-create.png)

> Tip: as you can see, the Terminal was changed so the nice icons are all displayed

The cluster has been successfully created, and because we are using Docker Desktop, the network is all set for us to use "as is".

So we can open the `Kubernetes master` URL in our Windows browser:

![kind-browser-k8s-master](/images/blog/2020-05-21-wsl2-dockerdesktop-k8s/wsl2-kind-browse-k8s-master.png)

And this is the real strength from Docker Desktop for Windows with the WSL2 backend. Docker really did an amazing integration.

## KinD: counting 1 - 2 - 3

Our first cluster was created and it's the "normal" one node cluster:

```bash
# Check how many nodes it created
kubectl get nodes
# Check the services for the whole cluster
kubectl get all --all-namespaces
```

![kind-list-nodes-services](/images/blog/2020-05-21-wsl2-dockerdesktop-k8s/wsl2-kind-list-nodes-services.png)

> While this will be enough for a lot of persons, let's leverage one of the coolest feature, multi-node clustering:

Reviewer 3: ```suggestion
While this will be enough for most people, let's leverage one of the coolest feature, multi-node clustering:

```

```bash
# Delete the existing cluster
kind delete cluster --name wslkind
# Create a config file for a 3 nodes cluster
cat << EOF > kind-3nodes.yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
  - role: control-plane
  - role: worker
  - role: worker
EOF
# Create a new cluster with the config file
kind create cluster --name wslkindmultinodes --config ./kind-3nodes.yaml
# Check how many nodes it created
kubectl get nodes
```

![kind-cluster-create-multinodes](/images/blog/2020-05-21-wsl2-dockerdesktop-k8s/wsl2-kind-cluster-create-multinodes.png)

> Tip: depending on how fast we run the "get nodes" command, it can be that not all the nodes are ready, wait few seconds and run it again, everything should be ready

> And that's it, we have created a three nodes cluster, and if we look at the services one more time, we will see several that have now three replicas:

Reviewer 3: ```suggestion
And that's it, we have created a three-node cluster, and if we look at the services one more time, we will see several that have now three replicas:

```

kubectl get all --all-namespaces
```

![wsl2-kind-list-services-multinodes](/images/blog/2020-05-21-wsl2-dockerdesktop-k8s/wsl2-kind-list-services-multinodes.png)

## KinD: can I see a nice dashboard?

Working on the command line is always good and very insightful. However, when dealing with Kubernetes we might want, at some point, to have a visual overview.

For that, the [Kubernetes Dashboard](https://github.com/kubernetes/dashboard) project has been created. The installation and first connection test is quite fast, so let's do it:

```bash
# Install the Dashboard application into our cluster
kubectl apply -f https://raw.githubusercontent.com/kubernetes/dashboard/v2.0.0-rc6/aio/deploy/recommended.yaml
# Check the resources it created based on the new namespace created
kubectl get all -n kubernetes-dashboard
```

![kind-install-dashboard](/images/blog/2020-05-21-wsl2-dockerdesktop-k8s/wsl2-kind-install-dashboard.png)

As it created a service with a ClusterIP (read: internal network address), we cannot reach it if we type the URL in our Windows browser:

![kind-browse-dashboard-error](/images/blog/2020-05-21-wsl2-dockerdesktop-k8s/wsl2-kind-browse-dashboard-error.png)

> That's why, if we continue reading the how-to, we need to create a temporary proxy:

Reviewer 3: ```suggestion
That's because we need to create a temporary proxy:

```

# Install the conntrack package
sudo apt install -y conntrack
```

![image-20200328174058258](/images/blog/2020-05-21-wsl2-dockerdesktop-k8s/wsl2-minikube-install conntrack.png)

Let's try to launch it again:

```bash
# Create a minikube one node cluster
minikube start --driver=none
# We got a permissions error > try again with sudo
sudo minikube start --driver=none
```

![image-20200328202611584](/images/blog/2020-05-21-wsl2-dockerdesktop-k8s/wsl2-minikube-start-error-systemd.png)

Ok, this error cloud be problematic ... in the past. Luckily for us, there's a solution

## Minikube: enabling SystemD

In order to enable SystemD on WSL2, we will apply the [scripts](https://forum.snapcraft.io/t/running-snaps-on-wsl2-insiders-only-for-now/13033) from [Daniel Llewellyn](https://twitter.com/diddledan).

I invite you to read the full blog post and how he came to the solution, and the various iterations he did to fix several issues.

So in a nutshell, here are the commands:

```bash
# Install the needed packages
sudo apt install -yqq daemonize dbus-user-session fontconfig
```

![image-20200328205328598](/images/blog/2020-05-21-wsl2-dockerdesktop-k8s/wsl2-minikube-systemd-packages.png)

> ```bash

Reviewer 2: This isn't quite a bash script. I think I'd separate the command to run `vi` from what the reader should paste in.
(also, personal taste, I'd use `sudoedit`).

Author: cc @nunix

Reviewer: oh ok, by scripts I mean the `/usr/sbin/enter-systemd-namespace` and `/usr/sbin/start-systemd-namespace`.
for the first part, it's indeed not a script 👍 

for the separation, while I agree, I still have to see something that "appeals" or make sense to me. By typing the first command, the reader will enter in `vi`, so not really sure how to make a separation here.

As for `visudo`, sorry my old Unix habits remained here 😄
